### PR TITLE
feature(docker-based-loaders): adding the design

### PR DESCRIPTION
**Pull request by fruch**
Adding the design for moving to all stress tools to be docker based

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: docs/docker-loaders.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Moving to docker based loaders
+
+### Current situation
+
+Currently, have fixed images for AWS, installing form repo/list in GCE/Azure
```
Comment (fgelcer) [inline docs/docker-loaders.md:5]: typo: `installing from`

Comment (fruch) [reply docs/docker-loaders.md]: the price of typing too fast....

Comment (fruch) [reply docs/docker-loaders.md]: fixed

**Review by fgelcer — DISMISSED** [review]
LGTM

File: docs/docker-loaders.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+             "cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"]
+```
+(other option would be to prepend the image, but I think it might be less clear)
+
+this is for cases we might have specific bugs in a nemesis for example that would prevent us from using latest scylla-bench, or latest cassandra-stress
+could be added once we encounter those specific requirements
+
+### Concerns
+
+* Preference tests might have slight different behaviors/timing when working under docker:
```
Comment (soyacz) [inline docs/docker-loaders.md:69]: do you mean performance tests?

File: docs/docker-loaders.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+for using multiple versions, we should be able to define the image as part of the stress command:
+
+```yaml
+# first would use version 4.6.5, and the 2nd would use the default
+stress_cmd: ["image=scylladb/scylla:4.6.5; cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"
+             "cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"]
+```
+(other option would be to prepend the image, but I think it might be less clear)
```
Comment (soyacz) [inline docs/docker-loaders.md:62]: your proposal looks ok, no need to add this comment.

**Review by soyacz — APPROVED** [review]
LGTM

File: docs/docker-loaders.md (hunk 2 above)
Comment (fruch) [reply docs/docker-loaders.md:69]: yes

Comment (vponomaryov) [reply docs/docker-loaders.md:69]: @fruch I think @soyacz also wanted to have the typo be fixed -> `s/Preference/Performance/`

Comment (fruch) [reply docs/docker-loaders.md:69]: Didn't thought otherwise :)

I won't merge before fixing, 

I was looking for more functional related comments/suggestion/concerns, 

care to help with a few words/details on how we'll use it on k8s  ?

File: docs/docker-loaders.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+could be added once we encounter those specific requirements
+
+### Concerns
+
+* Preference tests might have slight different behaviors/timing when working under docker:
+  we should strive to using the host networking, and pin enough CPUs to each stress threads running
+  we will need to run few candidates of the perf tests, with/without those changes, to evaluate.
+
+* for K8S setup we're already using the same image as the image under tests, but switch scylla-bench to docker for example,
+  would mean we'll need to refactor the way create loader pods (since in the current design, we'll end up with a docker running
```
Comment (vponomaryov) [inline docs/docker-loaders.md:74]: Running a docker service inside of a K8S pod is nonsense and
I don't see any reason to describe the situation exactly this way.

It must be something like following:
```
Since we will have separate images for different loader tools and
the need to run more than one during a test run
then we should have the possibility to create loaders of different types.
Having loaders of different types we will be able just to pick up those 
from dedicated python objects queues when needed.
```

File: docs/docker-loaders.md (hunk 2 above)
Comment (vponomaryov) [reply docs/docker-loaders.md:69]: Added below.

File: docs/docker-loaders.md (hunk 4 above)
Comment (fruch) [reply docs/docker-loaders.md:74]: > Running a docker service inside of a K8S pod is nonsense and I don't see any reason to describe the situation exactly this way.

We are not running docker service inside docker, but using docker client inside docker.
And I'm just explaining the current code/situation here. no that that's the goal or ideal thing to have.

Anyhow we have this issue also if we want to use the docker back-end, but I would be o.k. with docker backend have once solution, and K8S a different one

> 
> It must be something like following:
> 
> ```
> Since we will have separate images for different loader tools and
> the need to run more than one during a test run
> then we should have the possibility to create loaders of different types.
> Having loaders of different types we will be able just to pick up those 
> from dedicated python objects queues when needed.
> ```

**Comment by soyacz** [comment]
Question: when creating resources for loaders/monitor shall we provision AMI's with docker preinstalled or we should install it upon VM setup?

**Comment by fruch** [comment]
> Question: when creating resources for loaders/monitor shall we provision AMI's with docker preinstalled or we should install it upon VM setup?

For starter we'll install, but later we could switch into newer pre-installed image, only with docker. (and fallback into installing, for when we don't have pre-installed, similar to what we do today)

File: docs/docker-loaders.md (hunk 4 above)
Comment (vponomaryov) [reply docs/docker-loaders.md]: > @vponomaryov and how we would split the resources for each one ? example, we are using `cassandra-stress` for the main load, but during a nemesis we need `scylla-bench` how much resources we give `scylla-bench`, and how much we give cassandra stress ?
> 
> maybe we should put them in separated pods ? so we can spin them up as needed ?
> 
> in today's VM loader logic, we give each command all CPU (i.e. we don't limit), only if we use `stress_num` parameter of the loader thread, we pin down to specific CPU

Our current implementation for loaders in K8S is to use separate dedicated VM per each loader pod.
And each such pod has dedicated cores.
So, it is not "maybe we should put them in separated pods ?", it must be in separate pods only.
We can spin up loader pod dynamically if needed.
But now we create it during the provisioning stage and run there noop loop to make it be alive.
Then we just run loader commands on demand.

**Comment by vponomaryov** [comment]
Also, why do we want to stick to `docker`? Why not daemon-less `podman`?

**Comment by fruch** [comment]
> Also, why do we want to stick to `docker`? Why not daemon-less `podman`?

The way we build/publish images or how run, won't change

podman is much more problematic to use mostly for docker-inside-docker situations, we cause of hydra we need to support.

in general we don't any competing argument for using preferring podman over docker, 

take for example https://github.com/scylladb/scylla-dtest/issues/2729, in a much straight forward case of dtest, we run into rough edges, I'm not really know what to expect when we try much more complex usage of sct.

one more example, we are using the python package `docker`, we count of the docker demon socket/network api, podman can supply it, but it's still not fully complaint (at least last time I've checked)

long story short, it's big effort, since podman isn't exactly a drop in replacement.

File: docs/docker-loaders.md (hunk 4 above)
Comment (fruch) [reply docs/docker-loaders.md]: @vponomaryov I'm leaving to you, to explain what would be the design for k8s backends, for now it's left as it is.

i.e. for cassandra-stress, since we are using scylla image, we just skip the RemoteDocker part, and let it run on the node/pod. any other stress tool would be triggered as docker as currently implementation enabled it.

I'll need you to look a bit more into it explain the suggestion of how you think it should work, while keeping in mind we don't always know in advance which stress tools are going to be used.

File: docs/docker-loaders.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+For most of the tools we are going to create our docker images, under `scylladb/hydra-loaders` and with
+instructions on building each one in `/docker`, we should version them by the tool version if applicable, or by date.
+the only tool we are not going to build ourselves is cassandra-stress, which would be comping from the official docker images
+of scylladb.
+
+### Finding correct cassandra-stress image
+
+Since we are going to be using scylla docker images for loaders, we need a way to find the correct docker image, based on a
+given version or sha of scylla (both for OSS and Enterprise versions), if we'll find that match base on sha would be problematic
+we'll fall back to last available release (smaller than the currently under test version, ex. when testing 5.2-dev, we'll take 5.1.x version)
```
Comment (roydahan) [inline docs/docker-loaders.md:34]: In any case, we will of course need a way to control that (e.g when finding issue in new release and want to isolate the issue and make sure it's not due to a new c-s version).

File: docs/docker-loaders.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+stress_cmd: ["image=scylladb/scylla:4.6.5; cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"
+             "cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"]
+```
+
+this is for cases we might have specific bugs in a nemesis for example that would prevent us from using latest scylla-bench, or latest cassandra-stress
+could be added once we encounter those specific requirements
+
+### Concerns
+
+* Performance tests might have slight different behaviors/timing when working under docker:
```
Comment (roydahan) [inline docs/docker-loaders.md:68]: For the performance test themselves it's a less of concern because it'll take time for them to fully move to docker base (old branches, and when moving to new branch it's a new baseline).

I'm more concerned of the non-performance test which may change behaviour (e.g a longevity that should push enough load, or populate amount of information that will take more time).

I have an idea how we can test this quite easily.
@aleksbykov is working on a performance test that tries to find the "best throughput" and there the test tries to find an optimal usage in the amount of loaders, processes per loader and number of threads per process.
If we run the same with and without docker based c-s, we will be able to measure the differences and the limits.

File: docs/docker-loaders.md (hunk 5 above)
Comment (roydahan) [inline docs/docker-loaders.md:34]: We will need two versions of the image, one with scylla-driver and one with cassandra-driver.
It's needed at the moment for c-s, but we may need it also for other stress tools.

Comment (fruch) [reply docs/docker-loaders.md:34]: The official version is with the scylla-driver

If we will want cassandra-stress with cassandra-driver, we need to build a custom image specifically for that, from the source.

I'll create the instructions on how to create such image - but only for investigations, I don't think we should maintain a set of our images.

Also we should invest in disabling shard-aware at Scylla end, instead of using two versions all the time.

File: docs/docker-loaders.md (hunk 6 above)
Comment (fruch) [reply docs/docker-loaders.md:68]: Is that work ready to use already ?

Comment (roydahan) [reply docs/docker-loaders.md:68]: Yes, in terms of that we have the results of this big matrix.
It's not merged yet, but in final stages.
If you're ready from your side you can use Alex's commits.

File: docs/docker-loaders.md (hunk 5 above)
Comment (roydahan) [reply docs/docker-loaders.md:34]: 1. I think we do need to maintain it as part of our set of images.
The shard aware version has some limitations (e.g. the throughput is better without it).
Also, we would need to make sure that Scylla works well also without the shard-aware.

2. Even if inverting Scylla, we should have cases (that we don't have now) of mixed clients (shard aware and regular) and making sure that the regular client works well even when scylla side is not specially configured.

Comment (fruch) [reply docs/docker-loaders.md:34]: I've added explanation on how to build those,
https://github.com/scylladb/scylla-cluster-tests/pull/5146/commits/bb84cf7f36b29da7b3bab81aa742c5df4e278642

then you can put them in the configuration -
or inside the stress commands (Not implemented yet)
